src/float32sr.jl

```diff
@@ -83,6 +83,8 @@ end
 # Promotion
 Base.promote_rule(::Type{Float16}, ::Type{Float32sr}) = Float32
 Base.promote_rule(::Type{Float64}, ::Type{Float32sr}) = Float64
+Base.promote_rule(::Type{Float32}, ::Type{Float32sr}) = Float32
```

Comment:
I think I delibrately did not add this method as it could easily lead to cascading type instabilities. While the conversion is obviously allowed, `a*b` with `a` a `Float32` and `b` a `Float32sr` should have an explicit conversion.

Reply:
Ok, I am not sure I quite follow. Why is it ok to have a promotion rule for Float64 (L85) but not for Float32? Obviously, without this method the example given in the issue returns an error:

```ERROR: promotion of types Float32 and Float32sr failed to change any arguments``` 

since DE.jl internals do not know how to handle the promotion.

Do you have any suggestions here Milan?

Reply:
I agree, my comments have not been very consistent. At the moment in the main branch we have the following promotion logic

- `Base.promote_rule(::Type{FloatM}, ::Type{FloatNsr}) = FloatP` if `M != N` then `P = max(M,N)`

Meaning that 

1) we always promote to deterministic rounding
2) we always promote to the larger bitlength, but we
3) don't promote when the bitlengths match.

I still believe 1)+2) are good, but I agree that 3) feels weird given that other promotions are indeed defined. I probably had some cascading argument in mind, but I see that it's unfeasible to do and it's probably not the job of this package to guarantee type stability. So I suggest to define all the promotions between BFloat16, Float16/32/64 and their stochastic counterparts. Does that make sense to you?

Reply:
Yeah I think that is sensible. For the purposes of integration with DE.jl, it seems like the promotion rule is only required for `Float32sr`, not `Float16sr` or `Bfloat16sr` (e.g. https://github.com/tomkimpson/StochasticRounding.jl/blob/main/test/differential_equations.jl)

Reply:
Yes, but while we're at it may just include all of them?